src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -953,18 +963,26 @@ void RaftReplDev::handle_commit(repl_req_ptr_t rreq, bool recovery) {
 
 void RaftReplDev::handle_error(repl_req_ptr_t const& rreq, ReplServiceError err) {
     if (err == ReplServiceError::OK) { return; }
+    RD_LOGE("Raft Channel: Error in processing rreq=[{}] error={}", rreq->to_string(), err);
 
     if (!rreq->add_state_if_not_already(repl_req_state_t::ERRORED)) {
-        RD_LOGE("Raft Channel: Error in processing rreq=[{}] error={}", rreq->to_string(), err);
+        RD_LOGE("Raft Channel: Error has been added for rreq=[{}] error={}", rreq->to_string(), err);
         return;
     }
 
     // Remove from the map and thus its no longer accessible from applier_create_req
     m_repl_key_req_map.erase(rreq->rkey());
 
-    if (rreq->op_code() == journal_type_t::HS_DATA_INLINED) {
+    // Remove from the lsn map, so that it is no longer accessible from the state machine
+    // Since handle_error and append_log may occur concurrently,
+    // this removal is effective only if append_log is already processed
+    if (m_state_machine->lsn_to_req(rreq->lsn()) != nullptr) {
```

Comment:
>(m_state_machine->lsn_to_req(rreq->lsn()) != nullptr

this means the log has been appended to log store and pre_committed. if we want to discard this log, should we rollback it ?

Reply:
Actually I think the current scenario and issue are not very clear (it might require an emergent GC as mentioned in https://github.com/eBay/HomeStore/issues/607). The only scenario I can think of ATM is: the log is appended to the log store, but the data write fails, causing it to get stuck at the end_of_append_batch (and not get persisted). In this case, a rollback might not be necessary; instead, data should be retrieved from the data channel. Additionally, if a rollback is needed, I haven't yet figured out a suitable method to determine an appropriate rollback point.

Reply:
I would suggest lets carefully using `pre_commit_ext` in our current implementation,  only do whatever volatile.

The `pre_commit_ext` was called when log append to logstore,  (in our implementation it is memory only),  upon to this point,  (in our implementation) the log is not persisted to logstore, it is the `end_of_append_batch`,  ensures the persistency.   If a crash happened before `end_of_append_batch`,  the log can be volatilized from logstore,   no `rollback` will be called if the log get over-written.


Here,  we have to add the req to map(state-machine) to allow the `on_precommit` be executed,  but in fact the log is not persisted at all and no `commit` will be execute.

Better we differentiate "volatile log in SM" vs "non-volatile log in SM",  any log, after passing `end_of_append_batch` , it should NOT be touched from outside of SM as they can be committed already.

Technics can be done for example,
 
```
int64 ReplLogStore::to_volatile_lsn(uint64 lsn) {
  assert( lsn <= int64_max);
  return - int64_cast(lsn);
}

ReplLogStore::apend() {
  ......
  m_sm.link_lsn_to_req(rreq, to_volatile_lsn(lsn));
}
 ``` 
 
 ```
void ReplLogStore::end_of_append_batch(ulong start_lsn, ulong count) {
 ....
 for (int64_t lsn = int64_cast(start_lsn); lsn <= end_lsn; ++lsn) {
        auto volatile_lsn = to_volatile_lsn(lsn)
        auto rreq = m_sm.lsn_to_req(volatile_lsn);
        m_sm.link_lsn_to_req(rreq, lsn);
         m_sm.unlink_lsn_to_req(rreq, volatile_lsn);
        
 ```

Reply:
Agree that differentiating volatile vs non-volatile is a good approach as it clarifies the behavior of rreq in the state machine when encountering errors. For example, during error handling, if the rreq is volatile, we can perform an unlink, and if it is non-volatile, we might use ASSERT. However, I have some doubts about its necessity for the following reasons:
1. Currently, handle_error is expected to only deal with volatile logs, and non-volatile logs should not reach the error handling logic. Do we really need to introduce this concept at this stage, especially considering it reduces the available range of LSNs (even though our code defaults LSN to be int64 ATM)
2. For volatile logs, since we cannot control the order of link_lsn_to_req and failure of 'async write', when handle_error triggered, the rreq might or might not exist in the state machine. So we might still need to further discuss how to handle drive failures in the future. For example, should we check for the existence of volatile rreq in the state machine when triggering emergent gc, and if it exists, remove it? Otherwise, we may need to consider whether it is safe for link_lsn_to_req to overwrite volatile rreq.

Reply:
1.  The target is not trying to let non-volatile rreq reach handle_error but more on how to make the code more self-explain and easier to follow,  the feeling of uncertainty that a volatile rreq might be `unlink_lsn_to_req` is bad and every time we need to comfort ourselves there is no such case.

2. If a log is made to non-volatile at the end of `end_of_append_batch`, i.e passing the wait IO, even passing the log flush, for example L96 before `    sisl::VectorPool< repl_req_ptr_t >::free(reqs);`,  will that helps?

Reply:
uint64->int64 should be fine,  int64 can last for **2.92 Million Years** at 100K tps :)

Reply:
Updated. Since differentiating logs makes it easier to follow, I also update logic in `handle_error` and `link_lsn_to_req` to gather more information, especially for corner cases we might not have thought of. This change can resolve the conflict between 'append' and 'commit', but there's still an issue with the following scenario. So it's probably best not to unlink in `handle_error` and instead add some logs to check if an overwrite really happens in the state machine:

T1: Handle the raft event and pass.
T2: Append log to the log store and save the volatile LSN into the state machine.
T3: Fail to write due to an IO error, and then remove it from the state machine.
T4: In the pre-commit phase, can't find the volatile LSN, leading to a nullptr exception.

Reply:
agree not to do unlink atm.  unlink volatile lsn from SM doesnt help us., we can avoid putting error lsn as non-volatile into SM easily by checking rreq state `repl_req_state_t::ERRORED`.  Although we will not reaching the case before we have better disk write failure  handing

Reply:
Okay, added check for rreq state before convert lsn to non-volatile.